loader/src/references/reference_binder.py

```diff
@@ -0,0 +1,137 @@
+from dotty_dict import dotty
+
+from loader.src.config.logger import create_logger
+
+logger = create_logger("reference_binder")
+
+
+class ReferenceBinder:
+    def __init__(self, fhirstore):
+        self.fhirstore = fhirstore
+
+        # cache is a dict of form
+        # {
+        #   (fhir_type_source, value, system): {fhir_type_target: [fhir_id1, fhir_id2...]},
+        #   (fhir_type_source, value, system): {fhir_type_target: [fhir_id]},
+        #   ...
+        # }
+        self.cache = {}
+
+    def resolve_references(self, unresolved_fhir_object, reference_paths):
+        fhir_object = dotty(unresolved_fhir_object)
+
+        # iterate over the instance's references and try to resolve them
+        for reference_path in reference_paths:
+            logger.debug(
+                f"Trying to resolve reference for resource {fhir_object['id']}"
+                f"at {reference_path}"
+            )
+            try:
+                bound_ref = self.bind_existing_reference(fhir_object, reference_path)
+                fhir_object[reference_path] = bound_ref
+            except Exception as e:
+                logger.warning(
+                    "Error while binding reference for instance "
+                    f"{fhir_object} at path {reference_path}: {e}"
+                )
+
+        if fhir_object.get("identifier") and len(fhir_object["identifier"]) > 0:
+            self.resolve_pending_references(fhir_object)
+
+        return fhir_object.to_dict()
+
+    def bind_existing_reference(self, fhir_object, reference_path):
+        # FIXME: dotty-dict does not handle brackets indices,
+        # it uses dots instead (a.0.b instead of a[0].b)
+        reference_attribute = fhir_object[reference_path]
+
+        # If we have a list of references, we want to bind all of them.
+        # Thus, we loop on all the items in sub_fhir_object.
+        if not isinstance(reference_attribute, list):
+            reference_attribute = [reference_attribute]
+
+        for ref in reference_attribute:
+            # extract the type and itentifier of the reference
+            reference_type = ref["type"]
+            identifier = ref["identifier"]
+            if not identifier.get("value") or not identifier.get("system"):
```

Comment:
The syntax key in dict doesn't work with dotty?

Reply:
it does. changed.